<a href="https://colab.research.google.com/github/spinedace/AlgorithmsUN2023II/blob/main/Lab4/spinedaceLab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.238
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4400.47
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

5 7
12

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
9

Overwriting output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

4 7
11


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5
1 2 3 4 5
20


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

cc1plus: fatal error: max_pairwise_product_long_full.cpp: No such file or directory
compilation terminated.
./run_all_test.sh: line 5: input1.txt: No such file or directory
grep: output1.txt: No such file or directory
./run_all_test.sh: line 5: input2.txt: No such file or directory
grep: output2.txt: No such file or directory
./run_all_test.sh: line 5: input3.txt: No such file or directory
grep: output3.txt: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwis

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
3
5 100 300

Writing input4.txt


In [ ]:
100*300

30000

In [ ]:
%%writefile output4.txt
30000

Writing output4.txt


In [ ]:
%%writefile input5.txt
2
300 500

Writing input5.txt


In [ ]:
300*500

150000

In [ ]:
%%writefile output5.txt

Writing output5.txt


In [ ]:
%%writefile input6.txt
10
2 3 4 5 6 7 8 9 10 11

Writing input6.txt


In [ ]:
11*10

110

In [ ]:
%%writefile output6.txt
110

Writing output6.txt


In [ ]:
%%writefile input7.txt
4
200 150 20 300

In [ ]:
200*300

60000

In [ ]:
%%writefile output7.txt
60000

Writing output7.txt


In [ ]:
%%writefile input8.txt
5
10 9 8 7 6

Writing input8.txt


In [ ]:
10*9

90

In [ ]:
%%writefile output8.txt
90

Writing output8.txt


In [ ]:
%%writefile input9.txt
3
700 900 2

Writing input9.txt


In [ ]:
700*900

630000

In [ ]:
%%writefile output9.txt
630000

Writing output9.txt


In [ ]:
%%writefile input10.txt
7
2 3 5 7 11 13 17

Writing input10.txt


In [ ]:
13*17

221

In [ ]:
%%writefile output10.txt
221

Writing output10.txt


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..11}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

cc1plus: fatal error: max_pairwise_product_long_full.cpp: No such file or directory
compilation terminated.
./run_all_test.sh: line 5: input1.txt: No such file or directory
grep: output1.txt: No such file or directory
./run_all_test.sh: line 5: input2.txt: No such file or directory
grep: output2.txt: No such file or directory
./run_all_test.sh: line 5: input3.txt: No such file or directory
grep: output3.txt: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_long_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwis

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m55.301s
user	4m53.124s
sys	0m0.101s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m54.301s
user	4m52.421s
sys	0m0.070s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right) \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

571


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

44
44
601



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

572


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177092


In [ ]:
# Estimated time by instruction

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

# Estimate the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.859s
user	0m0.854s
sys	0m0.001s


## Python

In [ ]:
%%writefile MaxPairwiseProduct.py
def MaxPairwiseProduct(numbers):
    execints = 0  # N inst exec by line executed

    max_product = 0  # 1
    execints += 1

    aux_product = 0  # 1
    execints += 1

    n = len(numbers)  # 1
    execints += 1

    execints += 1
    for first in range(n):  # 2+2n
        execints += 2

        execints += 1
        for second in range(n):  # 2n + 2n**2
            execints += 2

            execints += 1
            if first != second:  # n**2
                aux_product = numbers[first] * numbers[second]  # n**2 - n
                execints += 1

                execints += 1
                if aux_product > max_product:  # n**2 - n
                    max_product = aux_product  # [1, n**2/2 - n/2] upper bound
                    execints += 1

    return max_product, execints  # 5 + 2*n + [5*n**2+1, 11*n**2/2 - n/2] upper bound

def main_function():
    execints = 0  # N inst exec by line executed

    n = int(input())  # 1
    execints += 1

    numbers = list(map(int, input().split()))  # n
    execints += n

    execints += 1

    execints += 1
    max_product, execints_returned = MaxPairwiseProduct(numbers)  # 1 + 1 + 4 + n + [5*n**2+1, 11*n**2/2 - n/2] upper bound
    execints += execints_returned

    execints += 1
    print(execints)   #11 + 6*n + [5*n**2+1, 11*n**2/2 - n/2]
    print(max_product)

if __name__ == '__main__':
    main_function()

Overwriting MaxPairwiseProduct.py


In [ ]:
!time python MaxPairwiseProduct.py < input13.txt

605137049
120989000

real	0m44.659s
user	0m44.275s
sys	0m0.035s


In [ ]:
# Estimated time by instruction
44/605137049

7.271080174765502e-08

(5.5n2+5.5n+11)×7.27∗10−**8s=1s

n≈1580

In [ ]:
# Verify input14.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input14.txt");

 MyFile << 1580 <<"\n";  // n
 for (int i=1; i <1580; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!time python MaxPairwiseProduct.py < input14.txt

12482815
2491662

real	0m0.947s
user	0m0.922s
sys	0m0.014s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Overwriting max_pairwise_product_long.cpp


In [ ]:
%%writefile max_pairwise_product_long_details.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execInstruction) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1
    execInstruction += 4;

    for (int first = 0; first < n; ++first) {                                //2+2n
        execInstruction += 2;
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
            execInstruction += 3;
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
                execInstruction++;
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    long long execInstruction = 0;
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    execInstruction += n;
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        execInstruction += 2;
        std::cin >> numbers[i];                           //n
        execInstruction++;
    }

    std::cout << MaxPairwiseProduct(numbers, execInstruction) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    std::cout << "Approximate executed instructions: " << execInstruction << std::endl;
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound



Overwriting max_pairwise_product_long_details.cpp


In [ ]:
!g++ max_pairwise_product_long_details.cpp -o max_pairwise_product_long_details
!./max_pairwise_product_long_details< input13.txt

120989000
Approximate executed instructions: 181664545


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long< input13.txt

120989000

real	0m0.410s
user	0m0.400s
sys	0m0.002s


In [ ]:
#Estimated time C++
0.41/181664545

2.2569070921351218e-09

Python


In [ ]:
%%writefile max_pairwise_product_long.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

if __name__ == '__main__':
    n = int(input())
    numbers = list(map(int, input().split()))

    result = max_pairwise_product(numbers)

    print(result)




Overwriting max_pairwise_product_long.py


In [ ]:
%%writefile max_pairwise_product_long_details.py
import time

def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)
    execInstruction = 4;

    for first in range(n):
        execInstruction += 2
        for second in range(first + 1, n):
            execInstruction += 3
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                execInstruction += 1
                max_product = aux_product

    return max_product, execInstruction

if __name__ == '__main__':
    execInstruction = 0
    n = int(input())
    execInstruction += 1
    numbers = list(map(int, input().split()))
    execInstruction += n
    execInstruction += 2 * n


    result, instructionCount = max_pairwise_product(numbers)



    print(result)

    print("Approximate executed instructions:", instructionCount)





Overwriting max_pairwise_product_long_details.py


In [ ]:
!time python max_pairwise_product_long.py < input14.txt

2491662

real	0m0.207s
user	0m0.195s
sys	0m0.010s


In [ ]:
!python max_pairwise_product_long_details.py < input14.txt

2491662
Approximate executed instructions: 3754098


In [ ]:
#Estimated time Python
0.20/3754098

5.327511428843893e-08

### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile sort_solution.cpp
/* A Program to sort the array using Introsort.
  The most popular C++ STL Algorithm- sort()
  uses Introsort. */

#include<bits/stdc++.h>
using namespace std;

// A utility function to swap the values pointed by
// the two pointers
void swapValue(int *a, int *b)
{
    int *temp = a;
    a = b;
    b = temp;
    return;
}

/* Function to sort an array using insertion sort*/
void InsertionSort(int arr[], int *begin, int *end)
{
    // Get the left and the right index of the subarray
    // to be sorted
    int left = begin - arr;
    int right = end - arr;

    for (int i = left+1; i <= right; i++)
    {
        int key = arr[i];
        int j = i-1;

       /* Move elements of arr[0..i-1], that are
          greater than key, to one position ahead
          of their current position */
        while (j >= left && arr[j] > key)
        {
            arr[j+1] = arr[j];
            j = j-1;
        }
        arr[j+1] = key;
   }

   return;
}

// A function to partition the array and return
// the partition point
int* Partition(int arr[], int low, int high)
{
    int pivot = arr[high];    // pivot
    int i = (low - 1);  // Index of smaller element

    for (int j = low; j <= high- 1; j++)
    {
        // If current element is smaller than or
        // equal to pivot
        if (arr[j] <= pivot)
        {
            // increment index of smaller element
            i++;

            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return (arr + i + 1);
}


// A function that find the middle of the
// values pointed by the pointers a, b, c
// and return that pointer
int *MedianOfThree(int * a, int * b, int * c)
{
    if (*a < *b && *b < *c)
        return (b);

    if (*a < *c && *c <= *b)
        return (c);

    if (*b <= *a && *a < *c)
        return (a);

    if (*b < *c && *c <= *a)
        return (c);

    if (*c <= *a && *a < *b)
        return (a);

    if (*c <= *b && *b <= *a)
        return (b);
}

// A Utility function to perform intro sort
void IntrosortUtil(int arr[], int * begin,
                  int * end, int depthLimit)
{
    // Count the number of elements
    int size = end - begin;

      // If partition size is low then do insertion sort
    if (size < 16)
    {
        InsertionSort(arr, begin, end);
        return;
    }

    // If the depth is zero use heapsort
    if (depthLimit == 0)
    {
        make_heap(begin, end+1);
        sort_heap(begin, end+1);
        return;
    }

    // Else use a median-of-three concept to
    // find a good pivot
    int * pivot = MedianOfThree(begin, begin+size/2, end);

    // Swap the values pointed by the two pointers
    swapValue(pivot, end);

   // Perform Quick Sort
    int * partitionPoint = Partition(arr, begin-arr, end-arr);
    IntrosortUtil(arr, begin, partitionPoint-1, depthLimit - 1);
    IntrosortUtil(arr, partitionPoint + 1, end, depthLimit - 1);

    return;
}

/* Implementation of introsort*/
void Introsort(int arr[], int *begin, int *end)
{
    int depthLimit = 2 * log(end-begin);

    // Perform a recursive Introsort
    IntrosortUtil(arr, begin, end, depthLimit);

      return;
}

// A utility function to print an array of size n
void printArray(int arr[], int n)
{
   for (int i=0; i < n; i++)
       printf("%d ", arr[i]);
   printf("\n");
}

// Driver program to test Introsort
int main()
{
    int n;
    std::cin >> n;
    int arr[n];
    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }

    // Pass the array, the pointer to the first element and
    // the pointer to the last element
    Introsort(arr, arr, arr+n-1);
    printArray(arr, n);

    return(0);
}

Overwriting sort_solution.cpp


Test with inputs 1 and 2



In [ ]:
!g++ sort_solution.cpp -o sort_solution_nlogn
!./sort_solution_nlogn < input1.txt
!./sort_solution_nlogn < input2.txt



sort_solution.cpp: In function ‘int* MedianOfThree(int*, int*, int*)’:
sort_solution.cpp:91:1: warning: control reaches end of non-void function []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wreturn-type-Wreturn-type]8;;]
   91 | }
      | ^
1 2 3 4 5 6 7 8 9 10 
1 2 3 4 5 6 7 8 9 10 


Test and time it with input 3

In [ ]:
!g++ sort_solution.cpp -o sort_solution_nlogn
!time ./sort_solution_nlogn < input3.txt


sort_solution.cpp: In function ‘int* MedianOfThree(int*, int*, int*)’:
sort_solution.cpp:91:1: warning: control reaches end of non-void function []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wreturn-type-Wreturn-type]8;;]
   91 | }
      | ^
1 3 9 23 200000 200000 

real	0m0.003s
user	0m0.001s
sys	0m0.001s


20 stress tests

In [ ]:
%%writefile sort_stress_test.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    srand(time(0)); // Inicializa la semilla del generador de números aleatorios con la hora actual

    for (int i = 20; i <= 40; i++) {
        string filename = "input" + to_string(i) + ".txt";
        ofstream outfile(filename);

        if (!outfile) {
            cerr << "Error al abrir el archivo " << filename << endl;
            return 1;
        }

        int numCount = i; // Cantidad de números en el archivo

        // Escribir la cantidad de números en la primera línea del archivo
        outfile << numCount << endl;

        // Genera números aleatorios y escribe en el archivo
        for (int j = 0; j < numCount; j++) {
            int num = rand() % 1000; // Números aleatorios entre 0 y 999 (puedes ajustar el rango según tus necesidades)
            outfile << num << " ";
        }

        outfile.close();
        cout << "Archivo " << filename << " creado exitosamente." << endl;
    }

    return 0;
}

Overwriting sort_stress_test.cpp


In [ ]:
!g++ sort_stress_test.cpp -o sort_stress_test
!./sort_stress_test

Archivo input20.txt creado exitosamente.
Archivo input21.txt creado exitosamente.
Archivo input22.txt creado exitosamente.
Archivo input23.txt creado exitosamente.
Archivo input24.txt creado exitosamente.
Archivo input25.txt creado exitosamente.
Archivo input26.txt creado exitosamente.
Archivo input27.txt creado exitosamente.
Archivo input28.txt creado exitosamente.
Archivo input29.txt creado exitosamente.
Archivo input30.txt creado exitosamente.
Archivo input31.txt creado exitosamente.
Archivo input32.txt creado exitosamente.
Archivo input33.txt creado exitosamente.
Archivo input34.txt creado exitosamente.
Archivo input35.txt creado exitosamente.
Archivo input36.txt creado exitosamente.
Archivo input37.txt creado exitosamente.
Archivo input38.txt creado exitosamente.
Archivo input39.txt creado exitosamente.
Archivo input40.txt creado exitosamente.


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
!g++ sort_solution.cpp -o sort_solution
!time ./sort_solution < input20.txt
!time ./sort_solution < input21.txt
!time ./sort_solution < input22.txt
!time ./sort_solution < input23.txt
!time ./sort_solution < input24.txt
!time ./sort_solution < input25.txt
!time ./sort_solution < input26.txt
!time ./sort_solution < input27.txt
!time ./sort_solution < input28.txt
!time ./sort_solution < input29.txt
!time ./sort_solution < input30.txt
!time ./sort_solution < input31.txt
!time ./sort_solution < input32.txt
!time ./sort_solution < input33.txt
!time ./sort_solution < input34.txt
!time ./sort_solution < input35.txt
!time ./sort_solution < input36.txt
!time ./sort_solution < input37.txt
!time ./sort_solution < input38.txt
!time ./sort_solution < input39.txt
!time ./sort_solution < input40.txt

sort_solution.cpp: In function ‘int* MedianOfThree(int*, int*, int*)’:
sort_solution.cpp:91:1: warning: control reaches end of non-void function []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wreturn-type-Wreturn-type]8;;]
   91 | }
      | ^
154 158 170 198 206 241 265 282 331 341 362 374 408 440 446 532 591 626 831 853 

real	0m0.003s
user	0m0.001s
sys	0m0.002s
120 177 248 288 350 380 410 420 491 494 565 610 683 727 751 764 828 857 874 907 973 

real	0m0.003s
user	0m0.002s
sys	0m0.000s
27 152 174 251 268 294 295 330 376 410 411 423 459 556 562 637 672 707 754 884 906 926 

real	0m0.003s
user	0m0.001s
sys	0m0.002s
40 46 70 133 189 220 234 364 479 506 540 543 622 653 656 681 693 787 881 892 923 930 998 

real	0m0.003s
user	0m0.001s
sys	0m0.001s
43 59 136 183 234 256 288 386 459 494 523 561 564 635 733 759 803 852 891 896 903 916 922 959 

real	0m0.003s
user	0m0.001s
sys	0m0.001s
66 77 163 176 239 240 269 307 373 388 453 514 526 528 670 674 679 717 721 763 826 85

Calculate estimated time by instruction and number of instructions by second

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile largest_number.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int n;
    cin >> n; // Lee la cantidad de números

    vector<int> lista(n);

    for (int i = 0; i < n; ++i) {
        cin >> lista[i]; // Lee los números y los almacena en el vector
    }

    int primerMax = 0; // Inicializa el primer máximo como el valor mínimo posible
    int segundoMax = 0; // Inicializa el segundo máximo como el valor mínimo posible

    for (int i = 0; i < n; ++i) {
        if (lista[i] > primerMax) {
            segundoMax = primerMax; // El antiguo primer máximo se convierte en el segundo máximo
            primerMax = lista[i]; // El nuevo primer máximo se actualiza
        } else if (lista[i] > segundoMax && lista[i] != primerMax) {
            segundoMax = lista[i]; // Actualiza el segundo máximo
        }
    }

    cout << "Primer número más grande: " << primerMax << endl;
    cout << "Segundo número más grande: " << segundoMax << endl;

    return 0;
}

Overwriting largest_number.cpp


Test with inputs 1 and 2



In [ ]:
!g++ largest_number.cpp -o largest_number
!./largest_number < input1.txt
!./largest_number < input2.txt

Primer número más grande: 10
Segundo número más grande: 9
Primer número más grande: 10
Segundo número más grande: 9


Test and time it with input 3

In [ ]:
!g++ largest_number.cpp -o largest_number
!./largest_number < input3.txt

Primer número más grande: 200000
Segundo número más grande: 23


20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3


Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ largest_number.cpp -o largest_number
!time ./largest_number < input20.txt
!time ./largest_number < input21.txt
!time ./largest_number < input22.txt
!time ./largest_number < input23.txt
!time ./largest_number < input24.txt
!time ./largest_number < input25.txt
!time ./largest_number < input26.txt
!time ./largest_number < input27.txt
!time ./largest_number < input28.txt
!time ./largest_number < input29.txt
!time ./largest_number < input30.txt
!time ./largest_number < input31.txt
!time ./largest_number < input32.txt
!time ./largest_number < input33.txt
!time ./largest_number < input34.txt
!time ./largest_number < input35.txt
!time ./largest_number < input36.txt
!time ./largest_number < input37.txt
!time ./largest_number < input38.txt
!time ./largest_number < input39.txt
!time ./largest_number < input40.txt

Primer número más grande: 853
Segundo número más grande: 831

real	0m0.003s
user	0m0.001s
sys	0m0.002s
Primer número más grande: 973
Segundo número más grande: 907

real	0m0.003s
user	0m0.003s
sys	0m0.000s
Primer número más grande: 926
Segundo número más grande: 906

real	0m0.004s
user	0m0.002s
sys	0m0.001s
Primer número más grande: 998
Segundo número más grande: 930

real	0m0.004s
user	0m0.003s
sys	0m0.000s
Primer número más grande: 959
Segundo número más grande: 922

real	0m0.003s
user	0m0.002s
sys	0m0.001s
Primer número más grande: 929
Segundo número más grande: 912

real	0m0.004s
user	0m0.000s
sys	0m0.003s
Primer número más grande: 958
Segundo número más grande: 872

real	0m0.003s
user	0m0.001s
sys	0m0.001s
Primer número más grande: 977
Segundo número más grande: 967

real	0m0.003s
user	0m0.001s
sys	0m0.002s
Primer número más grande: 976
Segundo número más grande: 945

real	0m0.003s
user	0m0.001s
sys	0m0.002s
Primer número más grande: 994
Segundo número más grande: 959

real	0m0.00

### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile largest_number_super.cpp
#include <iostream>

using namespace std;

int main() {
    int n;
    cin >> n; // Lee la cantidad de números

    int primerMax = 0; // Inicializa el primer máximo como el valor mínimo posible
    int segundoMax = 0; // Inicializa el segundo máximo como el valor mínimo posible

    for (int i = 0; i < n; ++i) {
        int numero;
        cin >> numero; // Lee un número

        if (numero > primerMax) {
            segundoMax = primerMax; // El antiguo primer máximo se convierte en el segundo máximo
            primerMax = numero; // El nuevo primer máximo se actualiza
        } else if (numero > segundoMax && numero != primerMax) {
            segundoMax = numero; // Actualiza el segundo máximo
        }
    }

    cout << "Primer número más grande: " << primerMax << endl;
    cout << "Segundo número más grande: " << segundoMax << endl;

    return 0;
}

Writing largest_number_super.cpp


Test with inputs 1 and 2




In [ ]:
!g++ largest_number_super.cpp -o largest_number_super
!./largest_number_super < input1.txt
!./largest_number_super < input2.txt

Primer número más grande: 10
Segundo número más grande: 9
Primer número más grande: 10
Segundo número más grande: 9


Test and time it with input 3

In [ ]:
!g++ largest_number_super.cpp -o largest_number_super
!./largest_number_super < input3.txt

Primer número más grande: 200000
Segundo número más grande: 23


20 stress tests

In [ ]:
!g++ largest_number_super.cpp -o largest_number_super
!time ./largest_number_super < input20.txt
!time ./largest_number_super < input21.txt
!time ./largest_number_super < input22.txt
!time ./largest_number_super < input23.txt
!time ./largest_number_super < input24.txt
!time ./largest_number_super < input25.txt
!time ./largest_number_super < input26.txt
!time ./largest_number_super < input27.txt
!time ./largest_number_super < input28.txt
!time ./largest_number_super < input29.txt
!time ./largest_number_super < input30.txt
!time ./largest_number_super < input31.txt
!time ./largest_number_super < input32.txt
!time ./largest_number_super < input33.txt
!time ./largest_number_super < input34.txt
!time ./largest_number_super < input35.txt
!time ./largest_number_super < input36.txt
!time ./largest_number_super < input37.txt
!time ./largest_number_super < input38.txt
!time ./largest_number_super < input39.txt
!time ./largest_number_super < input40.txt

Primer número más grande: 853
Segundo número más grande: 831

real	0m0.003s
user	0m0.002s
sys	0m0.000s
Primer número más grande: 973
Segundo número más grande: 907

real	0m0.003s
user	0m0.002s
sys	0m0.001s
Primer número más grande: 926
Segundo número más grande: 906

real	0m0.003s
user	0m0.003s
sys	0m0.000s
Primer número más grande: 998
Segundo número más grande: 930

real	0m0.003s
user	0m0.000s
sys	0m0.003s
Primer número más grande: 959
Segundo número más grande: 922

real	0m0.004s
user	0m0.002s
sys	0m0.000s
Primer número más grande: 929
Segundo número más grande: 912

real	0m0.003s
user	0m0.000s
sys	0m0.003s
Primer número más grande: 958
Segundo número más grande: 872

real	0m0.003s
user	0m0.001s
sys	0m0.001s
Primer número más grande: 977
Segundo número más grande: 967

real	0m0.004s
user	0m0.001s
sys	0m0.002s
Primer número más grande: 976
Segundo número más grande: 945

real	0m0.004s
user	0m0.000s
sys	0m0.003s
Primer número más grande: 994
Segundo número más grande: 959

real	0m0.00

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile largest_number_super.py
primer_max = segundo_max = float('-inf')

try:
    while True:
        numeros = input().split()  # Lee una línea y divide en números
        for numero in numeros:
            numero = int(numero)
            if numero > primer_max:
                segundo_max = primer_max
                primer_max = numero
            elif numero > segundo_max and numero != primer_max:
                segundo_max = numero

except EOFError:
    pass  # Maneja el EOF (End of File), en este punto se ha llegado al final de la entrada

print("Primer número más grande:", primer_max)
print("Segundo número más grande:", segundo_max)

Overwriting largest_number_super.py


Test with inputs 1 and 2




In [ ]:
!python largest_number_super.py < input1.txt
!python largest_number_super.py < input2.txt

Primer número más grande: 10
Segundo número más grande: 9
Primer número más grande: 10
Segundo número más grande: 9


Test and time it with input 3

In [ ]:
!python largest_number_super.py < input3.txt

Primer número más grande: 200000
Segundo número más grande: 23


20 stress tests

In [ ]:
!time python largest_number_super.py < input20.txt
!time python largest_number_super.py < input21.txt
!time python largest_number_super.py < input22.txt
!time python largest_number_super.py < input23.txt
!time python largest_number_super.py < input24.txt
!time python largest_number_super.py < input25.txt
!time python largest_number_super.py < input26.txt
!time python largest_number_super.py < input27.txt
!time python largest_number_super.py < input28.txt
!time python largest_number_super.py < input29.txt
!time python largest_number_super.py < input30.txt
!time python largest_number_super.py < input31.txt
!time python largest_number_super.py < input32.txt
!time python largest_number_super.py < input33.txt
!time python largest_number_super.py < input34.txt
!time python largest_number_super.py < input35.txt
!time python largest_number_super.py < input36.txt
!time python largest_number_super.py < input37.txt
!time python largest_number_super.py < input38.txt
!time python largest_number_super.py < input39.txt
!time python largest_number_super.py < input40.txt

Primer número más grande: 853
Segundo número más grande: 831

real	0m0.044s
user	0m0.035s
sys	0m0.008s
Primer número más grande: 973
Segundo número más grande: 907

real	0m0.046s
user	0m0.036s
sys	0m0.010s
Primer número más grande: 926
Segundo número más grande: 906

real	0m0.047s
user	0m0.035s
sys	0m0.010s
Primer número más grande: 998
Segundo número más grande: 930

real	0m0.059s
user	0m0.042s
sys	0m0.014s
Primer número más grande: 959
Segundo número más grande: 922

real	0m0.052s
user	0m0.036s
sys	0m0.009s
Primer número más grande: 929
Segundo número más grande: 912

real	0m0.049s
user	0m0.037s
sys	0m0.011s
Primer número más grande: 958
Segundo número más grande: 872

real	0m0.051s
user	0m0.036s
sys	0m0.011s
Primer número más grande: 977
Segundo número más grande: 967

real	0m0.045s
user	0m0.036s
sys	0m0.009s
Primer número más grande: 976
Segundo número más grande: 945

real	0m0.044s
user	0m0.034s
sys	0m0.010s
Primer número más grande: 994
Segundo número más grande: 959

real	0m0.04

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second